lib/common/http.ts

```diff
@@ -12,3 +12,98 @@
  * express or implied. See the License for the specific language governing
  * permissions and limitations under the License.
  */
+
+type HttpHeader = string[];
+
+export class HttpHeaders {
+    // Map from "header": [["HeAdEr", "value1"], ["HEADER", "value2"], ["header", "value3"]]
+    private headers: { [index: string]: [HttpHeader] } = {};
+
+    constructor(headers: HttpHeader[] = []) {
+        for (const header of headers) {
+            this.add(header[0], header[1]);
+        }
+    }
+
+    add(name: string, value: string) {
+        let values = this.headers[name.toLowerCase()];
+        if (values) {
+            values.push([name, value]);
+        } else {
+            this.set(name, value);
+        }
+    }
+
+    set(name: string, value: string) {
+        this.headers[name.toLowerCase()] = [[name, value]];
+    }
+
+    get_values(name: string) {
+        const values_list = this.headers[name.toLowerCase()] || [];
+        const values = [];
+        for (const entry of values_list) {
+            values.push(entry[1]);
+        }
+        return values;
+    }
+
+    get(name: string) {
+        const values = this.headers[name.toLowerCase()];
+        if (!values) { 
+            return "";
+        }
+        return values[0][1] || "";
+    }
+
+    remove(name: string) {
+        delete this.headers[name.toLowerCase()];
+    }
+
+    remove_value(name: string, value: string) {
+        let values = this.headers[name.toLowerCase()];
+        for (let idx = 0; idx < values.length; ++idx) {
+            const entry = values[idx];
+            if (entry[0] == name) {
+                if (values.length == 1) {
+                    delete this.headers[name.toLowerCase()];
+                } else {
+                    delete values[idx];
+                }
+            }
+        }
+    }
+
+    clear() {
+        this.headers = {};
+    }
+
+    // Allows for:
+    // let headers = new HttpHeaders();
+    // ...
+    // for (const header of headers) { }
+    *[Symbol.iterator]() {
+        for (const key in this.headers) {
+            const values = this.headers[key];
+            for (let entry of values) {
+                yield entry;
+            }
+        }
+    }
+
+    _flatten(): string[][] {
+        let flattened = [];
+        for (let key in this.headers) {
+            flattened.push([key, this.headers[key][0][0]]);
+        }
+        return flattened;
+    }
+}
+
+export class HttpRequest {
+    public headers = new HttpHeaders();
+    constructor(
+        public method: string,
+        public path: string,
+        public body?: string) {
```

Comment:
Should this be a stream of some kind?

Reply:
Eventually, yes. I'll come back and add stream support, this PR is big enough already.

Reply:
I also need to add support for custom request headers

Reply:
And then make elasticurl support those.